src/primitives/compression.h

```diff
@@ -0,0 +1,475 @@
+#include <consensus/amount.h>
+#include <script/script.h>
+#include <serialize.h>
+#include <uint256.h>
+#include <primitives/transaction.h>
+#include <script/standard.h>
+#include <cmath>
+#include <logging.h>
+#include <util/strencodings.h>
+#include <bitset>
+
+class CCompressedTxId
+{
+private:
+    uint32_t m_block_height;
+    uint32_t m_block_index;
+    uint256 m_txid;
+public:
+    const uint32_t&  block_height() const  { return m_block_height; }
+    const uint32_t& block_index() const { return m_block_index; }
+    const uint256& txid() const { return m_txid; }
+
+    explicit CCompressedTxId(const uint32_t& block_height, const uint32_t& block_index);
+    explicit CCompressedTxId(const uint256& txid);
+
+    bool IsCompressed() const {
+        return m_txid.IsNull();
+    }
+
+    friend bool operator==(const CCompressedTxId& a, const CCompressedTxId& b) {
+        return a.m_block_height == b.m_block_height && a.m_block_index == b.m_block_index && a.m_txid    == b.m_txid;
+    }
+    friend bool operator!=(const CCompressedTxId& a, const CCompressedTxId& b) {
+        return !(a == b);
+    }
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+        if (this->IsCompressed()) {
+            s << VARINT(this->m_block_height);
+            s << VARINT(this->m_block_index);
+        } else {
+            s.write(MakeByteSpan(this->m_txid));
+        }
+    }
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s, bool compressed) {
+        if (compressed) {
+            this->m_block_height = std::numeric_limits<uint32_t>::max();
+            this->m_block_index = std::numeric_limits<uint32_t>::max();
+            s >> VARINT(this->m_block_height);
+            s >> VARINT(this->m_block_index);
+        } else {
+            std::vector<unsigned char> txid(32);
+            s.read(MakeWritableByteSpan(txid));
+            this->m_txid = uint256(txid);
+        }
+    }
+
+    template <typename Stream>
+    CCompressedTxId(deserialize_type, Stream& s, bool compressed) {
+        Unserialize(s, compressed);
+    }
+
+    std::string ToString() const;
+};
+
+class CCompressedOutPoint
+{
+private:
+    CCompressedTxId m_txid;
+    uint32_t m_n;
+public:
+    const CCompressedTxId& txid() const { return m_txid; }
+    const uint32_t& n() const { return m_n; }
+
+    explicit CCompressedOutPoint(const CCompressedTxId& txid, const uint32_t& n);
+
+    friend bool operator==(const CCompressedOutPoint& a, const CCompressedOutPoint& b) {
+        return a.m_txid == b.m_txid && a.m_n == b.m_n;
+    }
+    friend bool operator!=(const CCompressedOutPoint& a, const CCompressedOutPoint& b) {
+        return !(a == b);
+    }
+
+    template <typename Stream>
+    inline void Serialize(Stream& s) const {
+        this->m_txid.Serialize(s);
+        s << VARINT(this->m_n);
+    }
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s, bool compressedTxId, bool constructor=false) {
```

Comment:
In 89b52782f8d6aca75d49048da1e137fbf615c0e6:

should always use `snake_case` for variable names, not camelCase.

Reply:
Solved, Thanks!